In [136]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [137]:
df = pd.read_csv(r"C:\Users\Krishna Murthy\Downloads\netflix.csv")
df

show_id     type                                        title  \
0     81145628    Movie      Norm of the North: King Sized Adventure   
1     80117401    Movie                   Jandino: Whatever it Takes   
2     70234439  TV Show                           Transformers Prime   
3     80058654  TV Show             Transformers: Robots in Disguise   
4     80125979    Movie                                 #realityhigh   
...        ...      ...                                          ...   
6231  80000063  TV Show                                 Red vs. Blue   
6232  70286564  TV Show                                        Maron   
6233  80116008    Movie       Little Baby Bum: Nursery Rhyme Friends   
6234  70281022  TV Show  A Young Doctor's Notebook and Other Stories   
6235  70153404  TV Show                                      Friends   

                      director  \
0     Richard Finn, Tim Maltby   
1                          NaN   
2                          NaN   
3                          NaN   
4             Fernando Lebrija   
...                        ...   
6231                       NaN   
6232                       NaN   
6233                       NaN   
6234                       NaN   
6235                       NaN   

                                                   cast  \
0     Alan Marriott, Andrew Toth, Brian Dobson, Cole...   
1                                      Jandino Asporaat   
2     Peter Cullen, Sumalee Montano, Frank Welker, J...   
3     Will Friedle, Darren Criss, Constance Zimmer, ...   
4     Nesta Cooper, Kate Walsh, John Michael Higgins...   
...                                                 ...   
6231  Burnie Burns, Jason Saldaña, Gustavo Sorola, ...   
6232  Marc Maron, Judd Hirsch, Josh Brener, Nora Zeh...   
6233                                                NaN   
6234  Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...   
6235  Jennifer Aniston, Courteney Cox, Lisa Kudrow, ...   

                                       country         date_added  \
0     United States, India, South Korea, China  September 9, 2019   
1                               United Kingdom  September 9, 2016   
2                                United States  September 8, 2018   
3                                United States  September 8, 2018   
4                                United States  September 8, 2017   
...                                        ...                ...   
6231                             United States                NaN   
6232                             United States                NaN   
6233                                       NaN                NaN   
6234                            United Kingdom                NaN   
6235                             United States                NaN   

     release_year    rating    duration  \
0            2019     TV-PG      90 min   
1            2016     TV-MA      94 min   
2            2013  TV-Y7-FV    1 Season   
3            2016     TV-Y7    1 Season   
4            2017     TV-14      99 min   
...           ...       ...         ...   
6231         2015        NR  13 Seasons   
6232         2016     TV-MA   4 Seasons   
6233         2016       NaN      60 min   
6234         2013     TV-MA   2 Seasons   
6235         2003     TV-14  10 Seasons   

                                              listed_in  \
0                    Children & Family Movies, Comedies   
1                                       Stand-Up Comedy   
2                                              Kids' TV   
3                                              Kids' TV   
4                                              Comedies   
...                                                 ...   
6231  TV Action & Adventure, TV Comedies, TV Sci-Fi ...   
6232                                        TV Comedies   
6233                                             Movies   
6234           British TV Shows, TV Comedies, TV Dramas   
6235                     Clas

In [138]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6236 entries, 0 to 6235
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       6236 non-null   object
 1   type          6235 non-null   object
 2   title         6234 non-null   object
 3   director      4265 non-null   object
 4   cast          5665 non-null   object
 5   country       5758 non-null   object
 6   date_added    6223 non-null   object
 7   release_year  6234 non-null   object
 8   rating        6224 non-null   object
 9   duration      6234 non-null   object
 10  listed_in     6233 non-null   object
 11  description   6233 non-null   object
dtypes: object(12)
memory usage: 584.8+ KB


# EDA

In [139]:
df.duplicated().sum()

0

In [140]:
# Replace null values with 'Unknown'
columns = ['country','director','cast','title','type','date_added','release_year','listed_in','duration','description','rating']
for col in columns:
    df[col] = df[col].fillna('Unknown')

In [141]:
df.isnull().sum()

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64

In [142]:
# Clean the 'duration' column (e.g., '1 season', '164 min')
df['duration_cleaned'] = df['duration'].apply(lambda x: str(x).replace(' Season', '').replace(' Seasons', '').replace(' min', '').strip())

# Convert the cleaned 'duration_cleaned' column to numeric
# Handle 'None' or invalid entries by setting them to NaN
df['duration_cleaned'] = pd.to_numeric(df['duration_cleaned'], errors='coerce')

# Fill missing values with the median or a default value (e.g., 0)
df['duration_cleaned'] = df['duration_cleaned'].fillna(df['duration_cleaned'].median())

In [143]:
df=df[df['type'].isin(['TV Show', 'Movie'])].copy()

In [144]:
# Convert numeric-like columns to numeric
df['release_year'] = pd.to_numeric(df['release_year'], errors='coerce')

In [148]:
df.head()

show_id     type                                    title  \
0  81145628    Movie  Norm of the North: King Sized Adventure   
1  80117401    Movie               Jandino: Whatever it Takes   
2  70234439  TV Show                       Transformers Prime   
3  80058654  TV Show         Transformers: Robots in Disguise   
4  80125979    Movie                             #realityhigh   

                   director  \
0  Richard Finn, Tim Maltby   
1                   Unknown   
2                   Unknown   
3                   Unknown   
4          Fernando Lebrija   

                                                cast  \
0  Alan Marriott, Andrew Toth, Brian Dobson, Cole...   
1                                   Jandino Asporaat   
2  Peter Cullen, Sumalee Montano, Frank Welker, J...   
3  Will Friedle, Darren Criss, Constance Zimmer, ...   
4  Nesta Cooper, Kate Walsh, John Michael Higgins...   

                                    country         date_added  release_year  \
0  United States, India, South Korea, China  September 9, 2019        2019.0   
1                            United Kingdom  September 9, 2016        2016.0   
2                             United States  September 8, 2018        2013.0   
3                             United States  September 8, 2018        2016.0   
4                             United States  September 8, 2017        2017.0   

     rating  duration                           listed_in  \
0     TV-PG    90 min  Children & Family Movies, Comedies   
1     TV-MA    94 min                     Stand-Up Comedy   
2  TV-Y7-FV  1 Season                            Kids' TV   
3     TV-Y7  1 Season                            Kids' TV   
4     TV-14    99 min                            Comedies   

                                         description  duration_cleaned  
0  Before planning an awesome wedding for his gra...              90.0  
1  Jandino Asporaat riffs on the challenges of ra...              94.0  
2  With the help of three human allies, the Autob...               1.0  
3  When a prison ship crash unleashes hundreds of...               1.0  
4  When nerdy high schooler Dani finally attracts...              99.0

# MODEL BUILDING

In [149]:
X = df[['duration_cleaned','rating','release_year']]
y = df['type']

In [150]:
# Splitting data
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,
                                                   test_size = 0.2,
                                                   random_state = 42)
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(4987, 3) (4987,)
(1247, 3) (1247,)


In [151]:
X_train.columns

Index(['duration_cleaned', 'rating', 'release_year'], dtype='object')

In [152]:
# Separating Categorical and Numerical Columns
X_train_num = X_train.select_dtypes(include=['float64'])
X_test_num = X_test.select_dtypes(include=['float64'])
print(X_train_num.shape)
print(X_test_num.shape)

(4987, 2)
(1247, 2)


In [153]:
X_train_num.columns

Index(['duration_cleaned', 'release_year'], dtype='object')

In [154]:
# Seperating Categorical columns
X_train_cat = X_train.select_dtypes(include=['object'])
X_test_cat = X_test.select_dtypes(include=['object'])
print(X_train_cat.shape)
print(X_test_cat.shape)

(4987, 1)
(1247, 1)


In [155]:
X_train_cat.columns

Index(['rating'], dtype='object')

In [156]:
X_train_cat.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4987 entries, 1025 to 860
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   rating  4987 non-null   object
dtypes: object(1)
memory usage: 77.9+ KB


## DATA PREPROCESSING

In [157]:
# Scaling the numerical data
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X_train_num_trans =sc.fit_transform(X_train_num)
X_test_num_trans =sc.transform(X_test_num)

In [158]:
print(X_train_num_trans.shape)
print(X_test_num_trans.shape)

(4987, 2)
(1247, 2)


In [159]:
#encoding categorical data
from sklearn.preprocessing import LabelEncoder

# Initialize the LabelEncoder
label_encoder = LabelEncoder()

# Encode the 'rating' column
X_train_rating_encoded = label_encoder.fit_transform(X_train_cat['rating']).reshape(-1,1)
X_test_rating_encoded = label_encoder.transform(X_test_cat['rating']).reshape(-1,1)

# Drop the original 'rating' column if not needed anymore
#X_train = X_train.drop('rating', axis=1)
#X_test = X_test.drop('rating', axis=1)

In [160]:
X_train_rating_encoded.shape

(4987, 1)

In [162]:
X_test_rating_encoded.shape

(1247, 1)

In [174]:
print(X_train_rating_encoded.shape)
print(X_test_rating_encoded.shape)

(4987, 1)
(1247, 1)


In [175]:
type(X_train_rating_encoded)

numpy.ndarray

In [179]:
X_train_transformed = np.concatenate((X_train_num_trans,X_train_rating_encoded), axis=1)
X_test_transformed =np.concatenate((X_test_num_trans,X_test_rating_encoded),axis=1)

In [180]:
type(X_train_transformed)

numpy.ndarray

In [181]:
X_train_transformed =pd.DataFrame(X_train_transformed)
X_test_transformed=pd.DataFrame(X_test_transformed)
print(type(X_train_transformed))
print(type(X_test_transformed))

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>


In [182]:
X_train_transformed

0         1    2
0     0.070894  0.192057  5.0
1    -0.146848  0.415532  8.0
2     0.245087  0.527270  8.0
3     0.266861  0.192057  9.0
4     1.050731  0.415532  6.0
...        ...       ...  ...
4982  0.353958  0.192057  5.0
4983  0.136216  0.303794  2.0
4984  0.680570  0.303794  9.0
4985  0.332184 -0.031419  5.0
4986 -1.671039  0.527270  6.0

[4987 rows x 3 columns]

In [183]:
X_test_transformed

0         1    2
0    -1.671039  0.080319  9.0
1     0.266861  0.303794  6.0
2    -1.671039  0.527270  8.0
3     0.223313  0.192057  6.0
4     0.593474  0.639008  6.0
...        ...       ...  ...
1242  0.571700 -1.372274  3.0
1243  1.181376 -2.824867  9.0
1244  0.266861  0.639008  8.0
1245 -1.671039 -0.813585  6.0
1246  0.484603  0.415532  8.0

[1247 rows x 3 columns]

In [184]:
from sklearn.preprocessing import LabelEncoder

# Initialize LabelEncoder
label_encoder = LabelEncoder()

# Fit and transform y_train and y_test
y_train_encoded = label_encoder.fit_transform(y_train)
y_test_encoded = label_encoder.transform(y_test)

# Now y_train_encoded and y_test_encoded will contain binary encoding (0 for 'movie', 1 for 'tv show')
print("y_train_encoded:", y_train_encoded)
print("y_test_encoded:", y_test_encoded)

y_train_encoded: [0 0 0 ... 0 0 1]
y_test_encoded: [1 1 1 ... 1 1 0]


# USING RANDOM FOREST CLASSIFIER

In [185]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Initialize RandomForestClassifier
rf_classifier = RandomForestClassifier(n_estimators=100,random_state=42)

In [205]:
# Train the model
rf_classifier.fit(X_train_transformed, y_train_encoded)

# Make predictions
y_pred = rf_classifier.predict(X_test_transformed)

In [206]:
# Evaluate the model
print("Accuracy on test data:", accuracy_score(y_test_encoded, y_pred))
print("Classification Report:")
print(classification_report(y_test_encoded, y_pred))
print("Confusion Matrix:")
print(confusion_matrix(y_test_encoded, y_pred))

Accuracy on test data: 0.9839615076182838
Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.98      0.99       832
           1       0.96      0.99      0.98       415

    accuracy                           0.98      1247
   macro avg       0.98      0.99      0.98      1247
weighted avg       0.98      0.98      0.98      1247

Confusion Matrix:
[[815  17]
 [  3 412]]


In [207]:
# Predict on the training data
y_train_pred = rf_classifier.predict(X_train_transformed)

# Calculate the accuracy on the training set
train_accuracy = accuracy_score(y_train_encoded, y_train_pred)

# Print the training accuracy
print("Training Accuracy: {:.2f}%".format(train_accuracy * 100))

Training Accuracy: 98.96%


## Key Observations:
### Training Accuracy: <b>98.96%
- The model is performing very well on the training data. It's able to fit the training set almost perfectly.

### Test Accuracy: <b>98.30%
- The model is also performing very well on the test data, which is close to the training accuracy. The difference between training and test accuracy is very small (only 0.4%), which indicates that the model is generalizing well to unseen data.

# HYPERPARAMETER TUNING

In [211]:
from sklearn.model_selection import GridSearchCV
import warnings
warnings.filterwarnings('ignore')
# Define the parameter grid for Random Forest
param_grid_rf = {
    'n_estimators': [50, 100, 200],           # Number of trees in the forest
    'max_depth': [10, 20, None],              # Maximum depth of the tree
    'min_samples_split': [2, 5, 10],          # Minimum number of samples required to split an internal node
    'min_samples_leaf': [1, 2, 4],            # Minimum number of samples required to be at a leaf node
    'bootstrap': [True, False]                # Whether bootstrap samples are used when building trees
}

# Initialize GridSearchCV for Random Forest
grid_search_rf = GridSearchCV(estimator=rf_classifier, param_grid=param_grid_rf, cv=3, n_jobs=-1, verbose=2)

# Fit GridSearchCV on the training data
grid_search_rf.fit(X_train_transformed, y_train_encoded)

# Get the best hyperparameters from the grid search
print("Best Hyperparameters for Random Forest: ", grid_search_rf.best_params_)

# Evaluate the model with the best hyperparameters
best_rf_model = grid_search_rf.best_estimator_

# Predict on the test set
y_test_pred_rf = best_rf_model.predict(X_test_transformed)

# Calculate the test accuracy
test_accuracy_rf = accuracy_score(y_test_encoded, y_test_pred_rf)
print("Test Accuracy (Random Forest) with Best Hyperparameters: {:.2f}%".format(test_accuracy_rf * 100))

Fitting 3 folds for each of 162 candidates, totalling 486 fits
Best Hyperparameters for Random Forest:  {'bootstrap': True, 'max_depth': 20, 'min_samples_leaf': 1, 'min_samples_split': 10, 'n_estimators': 100}
Test Accuracy (Random Forest) with Best Hyperparameters: 98.48%


# Conclusion:
### <b>Improved Generalization:
- *The slight improvement in test accuracy suggests that the model is now generalizing better to new, unseen data. The difference between training and test accuracy is still minimal (only 0.48%), which means the model is not overfitting and is performing well in terms of generalization.*

### <b>Better Model with Hyperparameter Tuning:
-  *The hyperparameter tuning has improved the model's performance on the test set without causing overfitting, which is a good sign that the model is well-optimized.*